Add FrameMemoryMonitor for LocalFrame GPU memory tracking during media loading

nullhook opened the pull request:
#### fd514e08ad086282a0fcb6076116d62664bc92d3
<pre>
Add FrameMemoryMonitor for LocalFrame GPU memory tracking during media loading
<a href="https://rdar.apple.com/150463891">rdar://150463891</a>

Reviewed by Brent Fulgham.

Introduce FrameMemoryMonitor to track LocalFrame GPU memory usage during media
loading. Reports memoryFootprint from RemoteMediaPlayerProxy to HTMLMediaElement
periodically via the progress timer. Enables unloading autoplay frames when
excessive GPU memory is detected.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
(WebCore::Document::frameMemoryMonitor):
(WebCore::Document::protectedFrameMemoryMonitor):
(WebCore::Document::parentFrameMemoryMonitorIfExists):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerDidReportGPUMemoryFootprint):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/loader/FrameMemoryMonitor.cpp: Added.
(WebCore::FrameMemoryMonitor::create):
(WebCore::FrameMemoryMonitor::FrameMemoryMonitor):
(WebCore::FrameMemoryMonitor::~FrameMemoryMonitor):
(WebCore::FrameMemoryMonitor::addUsage):
(WebCore::FrameMemoryMonitor::setMaxMemoryAllowedPerFrameForTesting):
(WebCore::FrameMemoryMonitor::checkMemoryPressureAndUnloadFrameIfNeeded):
(WebCore::FrameMemoryMonitor::parentFrameMemoryMonitorIfExists const):
* Source/WebCore/loader/FrameMemoryMonitor.h: Added.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::reportGPUMemoryFootprint const):
* Source/WebCore/platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerDidReportGPUMemoryFootprint):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::didLoadingProgress):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::reportGPUMemoryFootprint):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:

Canonical link: <a href="https://commits.webkit.org/294877@main">https://commits.webkit.org/294877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a9d29359c84c3944e7f56185c5051e8eff61153

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108510 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53980 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31564 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78534 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35486 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106343 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18107 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93223 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58867 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17944 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11264 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53337 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87745 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11326 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110885 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22502 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87534 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30840 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89420 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87171 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22196 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32028 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9747 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24809 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30404 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35723 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30210 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31772 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.cpp:
```diff
@@ -0,0 +1,96 @@
+
+namespace WebCore {
+
+Ref<FrameMemoryMonitor> FrameMemoryMonitor::create(LocalFrame& frame)
```
Why is this `LocalFrame&` rather than `const LocalFrame&`? Do we need to mutate the passed `frame` for some reason?

nullhook replied:
no, we don't need to mutate. switch to `const T&`

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.cpp:
```diff
@@ -0,0 +1,96 @@
+{
+}
+
+FrameMemoryMonitor::~FrameMemoryMonitor()
```
Maybe this could just be in the header as `~FrameMemoryMonitor() = default` ?

basuke replied:
Destructor is exported so the implementation should be here.

cdumez replied:
I agree with @brentfulgham . @basuke, Brent is asking to default the destructor in the cpp, not to inline it in the header.

nullhook replied:
fixed

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.cpp:
```diff
@@ -0,0 +1,96 @@
+        m_usageHasExceeded = true;
+
+        RefPtr frame = m_frame.get();
+
```
nit: Don't need this blank line here.

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.cpp:
```diff
@@ -0,0 +1,96 @@
+            return;
+
+        // FIXME: Prevent any navigations of an unloaded frame
+        frame->showResourceMonitoringError();
```
We would typically write this as:
```
if (RefPtr frame = m_frame.get())
    frame->showResourceMonitoringError();
```

basuke replied:
The original code of ResourceMonitor does more lines of code so I chose early return. In this case, I agree with Brent.

brentfulgham commented on Source/WebCore/platform/graphics/MediaPlayer.h:
```diff
@@ -286,13 +286,15 @@ class MediaPlayerClient : public CanMakeWeakPtr<MediaPlayerClient> {
 
     virtual Vector<RefPtr<PlatformTextTrack>> outOfBandTrackSources() { return { }; }
 
+
```
Nit: Please delete this extra line.

brentfulgham commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -347,6 +348,8 @@ void RemoteMediaPlayerProxy::setRate(double rate)
 {
     protectedPlayer()->didLoadingProgress(WTFMove(completionHandler));
+
+    protectedConnection()->send(Messages::MediaPlayerPrivateRemote::ReportGPUMemoryFootprint(WTF::memoryFootprint()), m_id);
```
How often do we hit this call when playing video? Will retrieving memory footprint every time be costly?

brentfulgham replied:
Looking through the code, it looks like we hit this every time an event timer fires, so it seems fine to do the memory check and send here.

nullhook replied:
> HTMLMediaElement::progressEventTimerFired() it’s a timer that fires every 250ms
> it already tracks how long it’s been since the last “progress”, i.e., when the MediaPlayer last reported that it loaded data

...seems like the right "tick" to report memory?

nullhook replied:
```cpp
 // 350ms is not magic, it is in the spec!
    m_progressEventTimer.startRepeating(350_ms);
```

correction: it's 350ms!

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.cpp:
```diff
@@ -0,0 +1,96 @@
+        return;
+
+    // FIXME: Add support for memory usage deltas
+    m_currentMemoryUsage = bytes;
```
This method name seems confusing -- It's really "setUsage" for now. Maybe we should use `setUsage` as the name until we have some mechanism for deltas?

nullhook replied:
hmm... the current logic can underreport sometimes and we report memory upwards, so it should always accumulate, which means we need HTMLMediaElement to manage these deltas?

since we're reporting process-level memory (rather than per-object), does the last frame win here? assuming there is an idle period between the frames below or memory is being freed between reports

Root Frame (FrameMemoryMonitor)
- Frame A (100 MB)
- Frame B (50 MB)

brentfulgham replied:
Does that mean that each `addMemory` message represents the current memory use of the media element? The way I thought this would work was that each media element uses some amount of memory. When we detect that one of the elements is using exceptional amounts of memory (perhaps for more than X measurements or something) that we would want to take action.

I'm not sure how holding a single value looking at all frames would be the right design for that. Shouldn't we be tracking it in some kind of table of "media element" to "memory use" or something?

nullhook replied:
`addUsage` (or `setUsage`) reflects the current memory snapshot from the GPUP. we're invoking it in a media player's tick, implicitly tying the memory usage to a specific media element.

the current limitation i'm concerned about is there's no attribution to which subframe caused the spike. so, when a threshold is exceeded at the root frame, we don't know at which depth the frame is responsible. iirc, there's a limit of 32 depth of frames.

the core issue is this:

```c++
if (RefPtr parentMonitor = parentFrameMemoryMonitorIfExists())
        parentMonitor->setUsage(bytes);
    else
        checkMemoryPressureAndUnloadFrameIfNeeded();
```

maybe we shouldn't propagate it upwards and just simply check memory pressure and unload?

nullhook replied:
ok, after carefully thinking about this i have removed the propagating logic. our model is different than RM's.

RM aggregates usage (deltas) in a bottom-up manner, FrameMemoryMonitor's model is self-contained, and the decision should be made at the same level.

basuke commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11599,6 +11601,29 @@ void Document::elementDisconnectedFromDocument(const Element& element)
 
+FrameMemoryMonitor& Document::frameMemoryMonitor()
+{
+    ASSERT(!frame()->isMainFrame());
```
I remember you said you need to monitor entire page's memory usage. If so, you need to have memoryMonitor for main frame then.

nullhook replied:
we will stick with subframes for now as we don't want to unload the main frame

basuke commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -4984,6 +4985,15 @@ void HTMLMediaElement::mediaPlayerDidRemoveVideoTrack(VideoTrackPrivate& track)
     track.willBeRemoved();
 }
 
+void HTMLMediaElement::mediaPlayerDidReportGPUMemoryFootprint(size_t footPrint)
```
Is this function called for increasing memory usage? Or is it reporting current memory footprint? The name implies current usage of memory. If so, adding usage in the following code seems weird.

nullhook replied:
it's reporting current memory footprint

basuke reviewed (commented):
What is the design of this monitor? Is this tracking the current usage of memory (which can be up and down), or total usage in the history of frame? This can be the big difference between this and ResourceMonitor.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11599,6 +11601,29 @@ void Document::elementDisconnectedFromDocument(const Element& element)
+    if (!m_frameMemoryMonitor)
+        m_frameMemoryMonitor = FrameMemoryMonitor::create(*frame());
+
+    return *m_frameMemoryMonitor.get();
```
`.get()` is unnecessary.

nullhook replied:
removed

basuke reviewed (changes requested):
Also note `void LocalFrame::showResourceMonitoringError()` is protected by `CONTENT_EXTENSION` which is not enabled in windows and playstation port. That's why some EWS fails.

Actually it reports the network usage limit of ResourceMonitor. I think you should extract the frame unloading part and create a new function for FrameMemoryMonitor to call that function.

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.cpp:
```diff
@@ -0,0 +1,90 @@
+        if (RefPtr document = m_frame->document()) {
+            if (document->quirks().shouldUnloadFrame()) {
+                // FIXME: Prevent any navigations of an unloaded frame
+                frame->showMemoryMonitorError();
```
This name is a little misleading -- it does more than just display an error -- it replaces the content of the frame with this message.

I propose you call this something like `replaceContentWIthMemoryMonitorError` or `unloadFrameAndShowMemoryMonitorError`.

nullhook replied:
makes sense. resolved!

brentfulgham commented on Source/WebCore/loader/FrameMemoryMonitor.h:
```diff
@@ -0,0 +1,58 @@
+    // FIXME: Add platform specific memory limits
+    size_t m_maxMemoryAllowedPerFrame = 1 * GB;
+    unsigned m_exceedCount = 0;
+    unsigned m_memoryPressureConsecutiveLimit = 3;
```
Nit: We usually initialize in modern style:

```
size_t m_maxMemoryAllowedPerFrame { 1 * GB };
unsigned m_exceedCount { 0 };
unsigned m_memoryPressureConsecutiveLimit { 3 };
```

brentfulgham reviewed (approved):
Looks good to me. Let's consider landing it if it shows as being performance neutral.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -912,6 +912,11 @@ bool Quirks::shouldEnableEnumerateDeviceQuirk() const
 
+bool Quirks::shouldUnloadFrame() const
+{
+    return needsQuirks();
```
This doesn't look right. `needsQuirks` just returns true unless the user has turned Quirks off entirely. This code means that we will always tell code checking `shouldUnloadFrame()` that we SHOULD unload the frame. That's not right.

brentfulgham replied:
This should do something like:

`return needsQuirks() && m_quirksData.shouldUnloadHeavyFrames;`

nullhook replied:
done

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2191,6 +2196,16 @@ static void handleWarbyParkerQuirks(QuirksData& quirksData, const URL& quirksURL
+
+    UNUSED_PARAM(quirksURL);
+    UNUSED_PARAM(documentURL);
+    UNUSED_PARAM(quirksData);
```
This should be setting a flag indicating that we need to do the unload. So I think you are missing a `bool` flag in QuirksData called something like, `shouldUnloadHeavyFrame` or something like that.

nullhook replied:
done

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -912,6 +912,11 @@ bool Quirks::shouldEnableEnumerateDeviceQuirk() const
 
+bool Quirks::shouldUnloadFrame() const
+{
+    return needsQuirks() & m_quirksData.shouldUnloadHeavyFrames;
```
Should use a logical `&&` here, not bitwise `&`.

brentfulgham replied:
Looks like the current version of the patch is still using a bitwise `&` rather than a boolean `&&`. Can you please fix that?

nullhook replied:
ah, this was a mistake. fixed

brentfulgham commented on Source/WebCore/page/Quirks.h:
```diff
@@ -145,6 +145,7 @@ class Quirks {
     bool shouldEnableSpeakerSelectionPermissionsPolicyQuirk() const;
     bool shouldEnableEnumerateDeviceQuirk() const;
 #endif
+    bool shouldUnloadFrame() const;
```
Please rename this to `shouldUnloadHeavyFrame()`

nullhook replied:
renamed

brentfulgham reviewed (changes requested):
I noticed two minor things that I'd like you to fix. Once those are corrected, I think this is ready to land!

brentfulgham reviewed (approved):
Looks great! Thank you for finishing those cleanups.

nullhook committed:
Add FrameMemoryMonitor for LocalFrame GPU memory tracking during media loading
rdar://150463891

Reviewed by Brent Fulgham.

Introduce FrameMemoryMonitor to track LocalFrame GPU memory usage during media
loading. Reports memoryFootprint from RemoteMediaPlayerProxy to HTMLMediaElement
periodically via the progress timer. Enables unloading autoplay frames when
excessive GPU memory is detected.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
(WebCore::Document::frameMemoryMonitor):
(WebCore::Document::protectedFrameMemoryMonitor):
(WebCore::Document::parentFrameMemoryMonitorIfExists):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerDidReportGPUMemoryFootprint):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/loader/FrameMemoryMonitor.cpp: Added.
(WebCore::FrameMemoryMonitor::create):
(WebCore::FrameMemoryMonitor::FrameMemoryMonitor):
(WebCore::FrameMemoryMonitor::~FrameMemoryMonitor):
(WebCore::FrameMemoryMonitor::addUsage):
(WebCore::FrameMemoryMonitor::setMaxMemoryAllowedPerFrameForTesting):
(WebCore::FrameMemoryMonitor::checkMemoryPressureAndUnloadFrameIfNeeded):
(WebCore::FrameMemoryMonitor::parentFrameMemoryMonitorIfExists const):
* Source/WebCore/loader/FrameMemoryMonitor.h: Added.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::reportGPUMemoryFootprint const):
* Source/WebCore/platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerDidReportGPUMemoryFootprint):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::didLoadingProgress):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::reportGPUMemoryFootprint):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:

Canonical link: https://commits.webkit.org/294877@main
